sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala

```diff
@@ -586,9 +584,139 @@ object ContainsExpressionBuilder extends StringBinaryPredicateExpressionBuilderB
 }
 
 case class Contains(left: Expression, right: Expression) extends StringPredicate {
-  override def compare(l: UTF8String, r: UTF8String): Boolean = l.contains(r)
+  override def inputTypes: Seq[DataType] = {
+    // TODO: properly handle String Types with differing collations
+    val leftIsString = left.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    val rightIsString = right.dataType.getClass.getName.startsWith(
+      "org.apache.spark.sql.types.StringType")
+    if (leftIsString && rightIsString) {
+      Seq(left.dataType, right.dataType)
+    } else {
+      Seq(StringType, StringType)
+    }
+  }
```

Comment:
What's the recommended way to resolve type matches?

When `left` and/or `right` Expressions are collated strings, their dataTypes are not equal. For example:

- when `left` is a (plain) string equal to "abc", its dataType is `StringType`
- when `right` is a collated string "C" using "UNICODE_CI", its dataType is `StringType(3)`

Although both types are essentially StringType, `override def inputTypes: Seq[DataType] = Seq(StringType, StringType)` doesn't work. What would be a preferred solution to go around this issue? Perhaps we should create a `CollatedStringType` that would match either `StringType` (without collation) or `StringType(#)` (with collation), and then override left & right expressions using Seq(CollatedStringType, CollatedStringType)?

Reply:
Can you try adding this line:
```scala
  override private[sql] def acceptsType(other: DataType): Boolean = other.isInstanceOf[StringType]
```

In `StringType.scala`? It's already part of OSS commits. I think that should take care of resolving type matching, without all the stuff that you are doing here.

Reply:
Starting off from https://github.com/dbatomic/spark/tree/collation_v2, I don't think I have that (err: Method 'acceptsType' overrides nothing). Let me try and pull some changes from master

Reply:
Update: works!